plugins/ccgx/fu-ccgx-i2c.c

```diff
@@ -0,0 +1,774 @@
+	w_value = ((scb_index << CY_SCB_INDEX_POS) | mode);
+	w_index = 0;
+	w_length = 0;
+
+	if (!g_usb_device_control_transfer (usb_device,
+					   G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_DEVICE,
+					   bm_request,
+					   w_value,
+					   w_index,
+					   NULL,
+					   w_length,
+					   &actual_length,
+					   io_timeout,
+					   NULL,
+					   &error_local)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "failed to reset i2c:%s", 
+			     error_local->message);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+i2c_get_status (FuDevice *device, CyI2CDeviceHandle* handle, guint8 mode, guint8 *i2c_status, GError **error)
+{
+	g_autoptr(GError) error_local =	NULL;
+	GUsbDevice *usb_device = NULL;
+	guint32 io_timeout = FU_CCGX_I2C_WAIT_TIMEOUT;
+	guint16 scb_index = 0;
+	guint16 w_value, w_index, w_length;
+	guint8 bm_request;
+	gsize actual_length = 0;
+
+	g_return_val_if_fail (handle != NULL, FALSE);
+	g_return_val_if_fail (FU_USB_DEVICE (device), FALSE);
+	usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	g_return_val_if_fail (usb_device != NULL, FALSE);
+
+	scb_index = handle->inf_num;
+
+	if (scb_index > 0)
+		scb_index = 1;
+
+	bm_request = CY_I2C_GET_STATUS_CMD;
+	w_value = ((scb_index << CY_SCB_INDEX_POS) | mode);
+	w_index = 0;
+	w_length = CY_I2C_GET_STATUS_LEN;
+	if (!g_usb_device_control_transfer (usb_device,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_DEVICE,
+					   bm_request,
+					   w_value,
+					   w_index,
+					   (guint8*)&i2c_status,
+					   w_length,
+					   &actual_length,
+					   io_timeout,
+					   NULL,
+					   &error_local)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "failed to get i2c status:%s", 
+			     error_local->message);
+
+		return FALSE;
+	}
+	return TRUE;
+
+}
+
+static gboolean
+wait_for_notification (FuDevice *device, CyI2CDeviceHandle* handle, guint16 *bytes_pending, guint32 io_timeout, GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+	GUsbDevice *usb_device = NULL;
+	gsize actual_length = 0;
+	guint8 ep_num;
+	guint8 i2c_status[CY_I2C_EVENT_NOTIFICATION_LEN] = {0};
+
+	g_return_val_if_fail (handle != NULL, FALSE);
+	g_return_val_if_fail (FU_USB_DEVICE (device), FALSE);
+	usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (device));
+	g_return_val_if_fail (usb_device != NULL, FALSE);
+
+	ep_num = handle->ep.intr_in;
+
+	if (g_usb_device_interrupt_transfer (usb_device,
+					     ep_num,
+					     (guint8*)i2c_status,
+					     CY_I2C_EVENT_NOTIFICATION_LEN,
+					     (gsize*)&actual_length,
+					     io_timeout,
+					     NULL,
+					     &error_local)) {
+
+		if (i2c_status[0] & CY_I2C_ERROR_BIT) {
+			if (i2c_status[0] & 0x80) { /* write */
+				if (i2c_reset (device,handle,CY_I2C_MODE_WRITE,NULL)  == FALSE) {
```

Comment:
if !(i2c_reset...) instead of check for false.
And isn't this a critical failure that you want to abort if the commands fail to send?  Or are you going to retry?

Reply:
If it isn't critical, you should use `g_autoptr(GError)` and then print the returned errors as a warning like `g_warning ("%s", error_local->message);`

Reply:
I will make i2c_reset can process GError.

Reply:
follow up question; this is the error path, and you're setting bytes_pending -- is that right?

Reply:
byte_pending is used for debugging in a previous USB driver.
byte_pending is not used in the current plugin.